c10/core/SymIntArrayRef.h

```diff
@@ -13,6 +13,337 @@
+  }
+
+  void debugCheckAnySymbolicInvariant() {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
+        AnySymbolic == computeSymbolic(),
+        "created SymIntArrayRef with incorrect AnySymbolic tag; real tag is ",
+        !AnySymbolic);
+  }
+
+  bool computeSymbolic() const {
+    for (const auto& s : *this) {
+      if (s.is_symbolic())
+        return true;
+    }
+    return false;
+  }
+
+ public:
+  /// @name Constructors
+  /// @{
+
+  /// Construct an empty ArrayRef.
+  /* implicit */ constexpr ArrayRef()
+      : Data(nullptr), Length(0), AnySymbolic(false) {}
+
+  /// Construct an ArrayRef from a single element.
+  ArrayRef(const T& OneElt)
+      : Data(&OneElt), Length(1), AnySymbolic(OneElt.is_symbolic()) {}
+
+  /// Construct an ArrayRef from a pointer and length.
+  ArrayRef(const T* data, size_t length)
+      : Data(data), Length(length), AnySymbolic(computeSymbolic()) {
+    debugCheckNullptrInvariant();
+  }
+
+  ArrayRef(Unchecked, const int64_t* data, size_t length)
+      : Data(reinterpret_cast<const c10::SymInt*>(data)),
+        Length(length),
+        AnySymbolic(false) {
+    debugCheckNullptrInvariant();
+    debugCheckAnySymbolicInvariant();
+  }
+
+  /// Construct an ArrayRef from a range.
+  ArrayRef(const T* begin, const T* end)
+      : Data(begin), Length(end - begin), AnySymbolic(computeSymbolic()) {
+    debugCheckNullptrInvariant();
+  }
+
+  /// Construct an ArrayRef from a SmallVector. This is templated in order to
+  /// avoid instantiating SmallVectorTemplateCommon<T> whenever we
+  /// copy-construct an ArrayRef.
+  template <typename U>
+  /* implicit */ ArrayRef(const SmallVectorTemplateCommon<T, U>& Vec)
+      : Data(Vec.data()), Length(Vec.size()), AnySymbolic(computeSymbolic()) {
+    debugCheckNullptrInvariant();
+  }
+
+  template <
+      typename Container,
+      typename = std::enable_if_t<std::is_same<
+          std::remove_const_t<decltype(std::declval<Container>().data())>,
+          T*>::value>>
+  /* implicit */ ArrayRef(const Container& container)
+      : Data(container.data()),
+        Length(container.size()),
+        AnySymbolic(computeSymbolic()) {
+    debugCheckNullptrInvariant();
+  }
+
+  /// Construct an ArrayRef from a std::vector.
+  // The enable_if stuff here makes sure that this isn't used for
+  // std::vector<bool>, because ArrayRef can't work on a std::vector<bool>
+  // bitfield.
+  template <typename A>
+  /* implicit */ ArrayRef(const std::vector<T, A>& Vec)
+      : Data(Vec.data()), Length(Vec.size()), AnySymbolic(computeSymbolic()) {
+    static_assert(
+        !std::is_same<T, bool>::value,
+        "ArrayRef<bool> cannot be constructed from a std::vector<bool> bitfield.");
+  }
+
+  /// Construct an ArrayRef from a std::array
+  template <size_t N>
+  /* implicit */ ArrayRef(const std::array<T, N>& Arr)
+      : Data(Arr.data()), Length(N), AnySymbolic(computeSymbolic()) {}
```

Comment:
Ho, std::vector<bool> is bad but std::array<bool> is ok?

Reply:
yes, because array of bool is not specialized to be bitfield (not really relevant here, but that's why)